### PR TITLE
Update pyright

### PR DESCRIPTION
**Pull request by srittau**

### CODE REVIEW

**Comment by srittau** [comment]
The main new warning is about argument disparities between class `IO` and other I/O classes: `IO` defines positional-or-named arguments, while other classes define positional-only arguments. For now, I suggest we just `# type: ignore` these warnings, but maybe we should have a discussion about making `IO` arguments positional-only, as it's a semi-protocol and many implementations expect positional-only arguments.

File: stdlib/unittest/mock.pyi [hunk 1]
```diff
@@ -406,7 +406,7 @@ if sys.version_info >= (3, 8):
         await_args_list: _CallList
     class AsyncMagicMixin(MagicMixin):
         def __init__(self, *args: Any, **kw: Any) -> None: ...
-    class AsyncMock(AsyncMockMixin, AsyncMagicMixin, Mock): ...
+    class AsyncMock(AsyncMockMixin, AsyncMagicMixin, Mock): ...  # type: ignore - argument disparities between base classes
```
Comment (AlexWaygood) [inline stdlib/unittest/mock.pyi:409]:
```suggestion
    class AsyncMock(AsyncMockMixin, AsyncMagicMixin, Mock): ...  # type: ignore # argument disparities between base classes
```

File: stdlib/io.pyi [hunk 2]
```diff
@@ -63,7 +63,7 @@ class BufferedIOBase(IOBase):
     def read(self, __size: int | None = ...) -> bytes: ...
     def read1(self, __size: int = ...) -> bytes: ...
 
-class FileIO(RawIOBase, BinaryIO):
+class FileIO(RawIOBase, BinaryIO):  # type: ignore - argument disparities between the base classes
```
Comment (AlexWaygood) [inline stdlib/io.pyi:66]:
```suggestion
class FileIO(RawIOBase, BinaryIO):  # type: ignore # argument disparities between the base classes
```

File: stdlib/io.pyi [hunk 3]
```diff
@@ -75,7 +75,7 @@ class FileIO(RawIOBase, BinaryIO):
     def read(self, __size: int = ...) -> bytes: ...
     def __enter__(self: Self) -> Self: ...
 
-class BytesIO(BufferedIOBase, BinaryIO):
+class BytesIO(BufferedIOBase, BinaryIO):  # type: ignore - argument disparities between the base classes
```
Comment (AlexWaygood) [inline stdlib/io.pyi:78]:
```suggestion
class BytesIO(BufferedIOBase, BinaryIO):  # type: ignore # argument disparities between the base classes
```

File: stdlib/io.pyi [hunk 4]
```diff
@@ -89,7 +89,7 @@ class BytesIO(BufferedIOBase, BinaryIO):
     else:
         def read1(self, __size: int | None) -> bytes: ...  # type: ignore[override]
 
-class BufferedReader(BufferedIOBase, BinaryIO):
+class BufferedReader(BufferedIOBase, BinaryIO):  # type: ignore - argument disparities between base classes
```
Comment (AlexWaygood) [inline stdlib/io.pyi:92]:
```suggestion
class BufferedReader(BufferedIOBase, BinaryIO):  # type: ignore # argument disparities between base classes
```

File: stdlib/io.pyi [hunk 5]
```diff
@@ -98,7 +98,7 @@ class BufferedReader(BufferedIOBase, BinaryIO):
     else:
         def read1(self, __size: int) -> bytes: ...  # type: ignore[override]
 
-class BufferedWriter(BufferedIOBase, BinaryIO):
+class BufferedWriter(BufferedIOBase, BinaryIO):  # type: ignore - argument disparities between base classes
```
Comment (AlexWaygood) [inline stdlib/io.pyi:101]:
```suggestion
class BufferedWriter(BufferedIOBase, BinaryIO):  # type: ignore # argument disparities between base classes
```

**Comment by srittau** [comment]
@erictraut pyright doesn't seem to like type aliases of generic types anymore. Is there a solution for this?

**Comment by AlexWaygood** [comment]
> @erictraut pyright doesn't seem to like type aliases of generic types anymore. Is there a solution for this?

I think this is the issue: https://github.com/microsoft/pyright/issues/2799#issuecomment-1005042304

**Comment by erictraut** [comment]
I recently fixed a bug where pyright was inconsistently interpreting type aliases that used a generic type without subscripts. By my interpretation, the type alias `X = List` is supposed to be interpreted as `X = List[Any]` (i.e. `X` is no longer generic). Is that consistent with your understanding? If not, is there any documentation that indicates the contrary?

**Comment by AlexWaygood** [comment]
@erictraut, that seems to be inconsistent to how it's treated at runtime, no?

```python
>>> d = dict[str, int]
>>> d[str, int]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: There are no type variables left in dict[str, int]
>>> e = dict
>>> e[str, int]
dict[str, int]
```

If `e = dict` were equivalent to `e = dict[Any, Any]`, then I would expect `e[str, int]` to fail at runtime, just as `d[str, int]` does.

**Comment by erictraut** [comment]
It seems to me we have several options for type aliases that use generic classes (or other generic type aliases) in their definitions but omit type arguments:
1. Treat it not as a type alias but as a variable definition, unless it it explicitly annotated as a `TypeAlias`, in which case the missing type aliases are treated as `Any`.
2. Treat it as a type alias and assume the missing type arguments are meant to be `Any`. This is consistent with how generic types and type aliases are used everywhere else.
3. Treat it as a type alias but assume the missing type arguments are meant to remain unpopulated type parameters, and the resulting type alias should be generic.

PEP 484 is not clear on which interpretation is correct, although it does say that "Type aliases may be as complex as type hints in annotations -- anything that is acceptable as a type hint is acceptable in a type alias", which seems to imply that the RHS of a type alias should be interpreted the same as if that type expression were used in any other annotation. That would imply that missing type arguments should be treated as `Any` (option 2).

I think you're arguing for option 3. This is not only inconsistent with the rules of generic types everywhere else, but it also creates challenging ambiguities. For example, consider the type alias `X = dict | OrderedDict`. If we use interpretation 3, this would be a generic type alias, but does it have four type parameters? Two? If it's two, on what basis would a type checker know to combine the type parameters from `dict` and `OrderedDict`? Is it based on the name of the type parameters that happen to be used in those class declarations?

My assumption is that option 2 is the correct interpretation, but I think option 1 is also reasonable given your argument about runtime compatibility. But I suspect option 1 would break quite a bit of assumptions in existing code.

For either option 1 or 2, the typeshed stubs would need to be fixed.

I'll add another option:
4. Define a special-case rule for an implicit type aliases where the RHS consists of a single generic class (or type alias) where the type arguments are omitted. In this specific case, the generic type arguments would be treated as though they were present and copied from the generic class (or type alias) used in the RHS. If the RHS uses any subscripts or unions, it would no longer fall under the special case, and all missing type arguments would be treated as `Any`.

Option 4 might be the least disruptive and the most intuitive, but it's also rather complicated and inconsistent.

Thoughts? Perhaps we should move this discussion to the python/typing forum to get additional opinions and perspectives?

**Comment by srittau** [comment]
I am fine with either interpretation, although intuitively to me an alias would include the generic arguments of the original, not interpret it as `Any`. But I agree with Eric that this discussion should be moved somewhere else.

For typeshed I don't mind to use more explicit annotations, explicit is better than implicit, after all.

**Comment by AlexWaygood** [comment]
Yeah, for clarity: I'm not arguing either way really. It may be that the runtime behaviour needs to change! And I agree with @srittau that explicit is better than implicit.

But I do think the difference in behaviour between the runtime and the type-checking needs to be resolved.

**Comment by AlexWaygood** [comment]
And yes, this discussion should be moved to python/typing.

**Comment by erictraut** [comment]
OK, thanks. I will start a discussion thread on python/typing.

**Comment by erictraut** [comment]
https://github.com/python/typing/discussions/1011

**Comment by vrslev** [comment]
Should I close this PR? https://github.com/python/typeshed/pull/6817

File: stdlib/bz2.pyi [hunk 6]
```diff
@@ -88,7 +88,7 @@ def open(
     newline: str | None = ...,
 ) -> BZ2File | TextIO: ...
 
-class BZ2File(BaseStream, IO[bytes]):
+class BZ2File(BaseStream, IO[bytes]):  # type: ignore # argument disparities between base classes
```
Comment (JelleZijlstra) [inline stdlib/bz2.pyi:91]: Agree with a comment above that we should make IO params pos-only, but this is fine for now.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by erictraut** [comment]
I just published pyright 1.1.205, which incorporates option 4 above. That seems to be consistent with mypy's behavior and is the consensus from people who responded to the python/typing thread.

**Pull request merged**